framework/src/org/checkerframework/framework/util/MultiGraphQualifierHierarchy.java

```diff
@@ -717,6 +717,10 @@ assert getTopAnnotation(a1) == getTopAnnotation(a2) :
         }
         return newglbs;
     }
+    
+    protected boolean annotationMirrorsAreEqual(AnnotationMirror a1, AnnotationMirror a2) {
```

Comment:
A comment about point 2 regarding aliases: AnnotatedTypeMirror.addAnnotation substitutes all aliases. After that, no aliases should matter. As long as you construct AnnotatedTypeMirrors with your annotations and then compute their GLB, there is no issue with aliases.
The problem is with methods that directly accept AnnotationMirrors, which might refer to aliases, e.g. if they come from a Tree. We should extract that alias management code from AnnotatedTypeMirror and make sure that AnnotationMirrors everywhere pass through canonization.